# Lab: Hypothesis Tests 📔❔
In this lab, you will conduct hypothesis tests on the Hello Dataset! 

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab TA, your CAs, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

In [ ]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Alex"
ca1_name = "Ana"
ca2_name = "Evan"


# Say hello to each other!
# - Groups of 3 are ideal :)
# - However, groups of 2 or 4 are fine too!

# QOTD to ask your group: What is your favorite holiday?
partner1_name = ""
partner1_netid = ""
partner1_favholiday = ""

partner2_name = ""
partner2_netid = ""
partner2_favholiday = ""

partner3_name = ""
partner3_netid = ""
partner3_favholiday = ""

<hr style="color: #DD3403;">

# Part 1: Sleep
Remember the Hello Survey you took towards the beginning of the semester? Well, we are bringing it back! We're going to conduct hypothesis tests using **your Hello Dataset**. Specifically, we will answer this question: 

> **Do UIUC students get significantly different sleep compared to the average American?**

First, as always, we must import the Hello Dataset. 

The "Hello Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/hello-fa23.csv
```

Use Python to load this dataset into a DataFrame called `df_hello`:

In [3]:
import pandas as pd
df_hello = pd.read_csv("https://waf.cs.illinois.edu/discovery/hello-fa23.csv")
df_hello

Name   Section  \
0        Prof. Karle    L1, L2   
1         Prof. Wade    L1, L2   
2             Andrew       Y05   
3           Annabel        Y05   
4             Ayden        Y14   
...              ...       ...   
1069         CA Soph  Y07, Y16   
1070          CA Ana       Y14   
1071  CA Sri Nithya        Y26   
1072       CA Lauren      YALL   
1073        CA Subin       Y05   

     What is your first name you want us to use for you?  \
0                                           Prof. Karle    
1                                            Prof. Wade    
2                                                Andrew    
3                                              Annabel     
4                                                Ayden     
...                                                 ...    
1069                                            CA Soph    
1070                                             CA Ana    
1071                                     CA Sri Nithya     
1072                                          CA Lauren    
1073                                           CA Subin    

     What lab section are you enrolled in?  (Use the three-letter code that starts with a "Y", like "Y42".)  \
0                                                L1, L2                                                       
1                                                L1, L2                                                       
2                                                   Y05                                                       
3                                                   Y05                                                       
4                                                   Y14                                                       
...                                                 ...                                                       
1069                                           Y07, Y16                                                       
1070                                                Y14                                                       
1071                                                Y26                                                       
1072                                               YALL                                                       
1073                                                Y05                                                       

     Does your NetID ending with odd number or even number?  \
0                         NetID ends with an odd number       
1                                                   NaN       
2                         NetID ends with an odd number       
3                        NetID ends with an even number       
4                        NetID ends with an even number       
...                                                 ...       
1069                     NetID ends with an even number       
1070                     NetID ends with an even number       
1071                     NetID ends with an even number       
1072                     NetID ends with an even number       
1073                     NetID ends with an even number       

     If I flipped a coin, do you think it would it land on heads or tails?  \
0                                                 Heads                      
1                                                 Heads                      
2                                                 Heads                      
3                                                 Heads                      
4                                                 Heads                      
...                                                 ...                      
1069                                              Tails                      
1070                                              Heads                      
1071                                              Heads                      
1072            

### 🔬 Test Case Checkpoint 🔬

In [4]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED 373all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert("df_hello" in vars()), "This is not the Hello dataset you're looking for. Check the URL."
assert("Do you like ice cream?" in df_hello), "This is not the Hello dataset you're looking for. Check the URL."
assert(len(df_hello) == 1074), "This is not the Hello dataset you're looking for. Check the URL."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Setting Up The Test
In 2013, [Gallup completed their latest survey on sleep](https://news.gallup.com/poll/166553/less-recommended-amount-sleep.aspx) that shows that the average amount of sleep Americans get is **6.8 hours**.  

To determine whether UIUC students get significantly more or less sleep than the average American, we will treat the **FALL 23 students in DISCOVERY** as a **random sample** of the student population at UIUC. Then, we will conduct a **one-sample, two-sided z-test** to answer our question. 

Before we begin, we must define our hypotheses. Remember, the average amount of sleep found in Gallup's study is **6.8** hours, and we want to determine if our random sample is **significantly different** from this average. 

**Q1: Write your null and alternative hypothesis using the cell below. Hint: you may find the following symbol useful for the alternative hypothesis: $\neq$** (double-click on this cell to copy it)

Null Hypothesis: The observed difference between the average American 6.8 hours of sleep and the DISCOVERY sample is due to chance.

Alternative Hypothesis: The observed difference between the average American 6.8 hours of sleep and the DISCOVERY sample is due something other than chance.

## Puzzle 1.1: Summary Statistics
Now that we've defined our hypotheses, we will need to find some basic **summary statistics** about the **average hours of sleep** of students in DISCOVERY before we conduct our **z-test**. Our `df_hello` contains the average hours of sleep of students in DISCOVERY under the column named "How many hours do you sleep in a day?".

Assign this column name to a **string variable** named `sleep_question` using the cell below:

In [9]:
sleep_question = "How many hours do you sleep in a day?"
sleep_question

'How many hours do you sleep in a day?'

Now, using `sleep_question` and your `df_hello`, find the **mean hours of sleep** of everyone in DISCOVERY, storing the result in the **variable** `sleep_mean`:

In [10]:
sleep_mean = df_hello[sleep_question].mean()
sleep_mean

7.446354663801338

Next, find the **standard deviation** of hours of sleep of everyone in DISCOVERY, storing the result in the **variable** `sleep_std`:

In [11]:
sleep_std = df_hello[sleep_question].std()
sleep_std

3.1705576760568746

Finally, find the **sample size** we're using for this z-test, storing the result in the **variable** `sample_size`. 

There's something a little tricky about the data we have in this column. Some people haven't filled out the answer to this captivating question, causing `NaN` (empty) values to appear. Dealing with **null values** is a **common occurrence** in data science. To exclude empty values when finding the sample size, the `.count()` function will be useful, as it excludes cells with NaN values when counting columns or rows.

In [14]:
sample_size = df_hello[sleep_question].count()
sample_size

1047

### 🔬 Test Case Checkpoint 🔬

In [15]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
import math
names = ['sleep_question', 'sleep_mean', 'sleep_std', 'sample_size']
for name in names:
    assert(name in vars()), "Double check that your variables are named correctly."
assert(math.isclose(df_hello[sleep_question].var(), 10.052435977203169)),"Double check the column question text you've defined in 'sleep_question'." 
assert(math.isclose(sleep_mean, 7.446354663801338)), "Your 'sleep_mean' is incorrect."
assert(math.isclose(sleep_std, 3.1705576760568746)), "Your 'sleep_std' is incorrect."
assert(sample_size**2 / 8 == 137026.125), "Your 'sample_size' is incorrect. Make sure you are excluding empty values when finding the sample size."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Puzzle 1.2: Finding our P-value
Now that we've **calculated and stored** the necessary statistics, we can begin calculating the **p-value** and drawing conclusions about our hypotheses. For the purposes of this lab, we will utilize a significance level of $0.05$. 

Because we are doing a two-tailed z-test, when we find the p-value, we want to make sure to include both tails. An easy way to do this is to find one tail and multiply it by 2 :) 

To begin, **run the following import statement**: 

In [16]:
import scipy.stats


This imports the `scipy.stats` library containing functions necessary to calculate our complex statistics and normal distribution values. 

In the case that the `scipy` module is not found:
- If you are on Windows, open the **terminal** and run `py -m pip install scipy`
- If you are on Mac (OS X): open the **terminal** and run  `python3 -m pip install scipy`

If you are still having import errors, let your lab CAs and TAs know! They will help you install the necessary libraries. 


Now, calculate the **standard error** in the hours of sleep, storing the result in the **numeric variable** `standard_error`. 

Remember: $SE = \frac{SD}{\sqrt{n}}$, where $SE$ is standard error, $SD$ is standard deviation, and $n$ is the sample size. 

In [17]:
standard_error = sleep_std / sample_size**(1/2)
standard_error

0.09798561370928813

Next, **calculate our test statistic** given the `sleep_mean` and `standard_error` found above, storing the result in the **numeric variable** `z_stat`. 

Remember: $z = \frac{\text{value - EV}}{\text{SE}}$, where $z$ is the z-statistic, $\text{value}$ is the mean of the data, $\text{EV}$ is the expected value, and $\text{SE}$ is the standard error. 

In [27]:
z_stat = (sleep_mean - 6.8) / standard_error
z_stat

6.596424100776638

Finally, we can calculate our p-value! Find it and store our p-value in the **variable** `p_value`.

This would be the area to the **left or right** of our `z-stat` on the **normal distribution**. 

Remember that we are are doing a **two-sided** z-test. Because of this, when we find the p-value, we want to make sure to include both tails. An easy way to do this is to find one tail and multiply it by 2 :) 

In [30]:
D = scipy.stats.norm()
p_value = 2 * (1 - D.cdf(z_stat))
p_value

4.211919701901934e-11

Remember, the p-value is the probability of getting the data we have in DISCOVERY's `df_hello` or something even more extreme **if the null were true**. We default to **"accepting"** the **null** hypothesis if the p-value is **larger than** our threshold of $0.05$, although statisticians prefer we say **do not reject** the null instead of accept.

### 🔬 Test Case Checkpoint 🔬

In [31]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
names = ['standard_error', 'z_stat', 'p_value']
for name in names:
    assert(name in vars()), "Double check that your variables are named correctly."
assert(math.isclose(standard_error, 0.09798561370928813)), "Your 'standard_error' is incorrect."
assert(math.isclose(z_stat, 6.596424100776638)), "Your 'z_stat' is incorrect."
assert(math.isclose(p_value, 4.211919701901934e-11, abs_tol=1e-15)), "Your 'p_value' is incorrect. Try and visualize the normal distribution and the areas to the left or right of your z-stat."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Interpretation 
**Q2: We have now observed the p-value for our hypothesis test. Do we reject the null hypothesis? Explain why or why not. Explain what this means in the context of our original question: Do Illinois students get significantly different sleep compared to the average American? Make sure you answer *both* questions.** 

Remember, our significance level is $0.05$.



p = 4.21 x 10^-11, which is less than 0.05 so we reject the null hypothesis. This means that Illinois students get significantly different sleep compared to the average American.

## Doing it in Python
We had to do a significant amount of calculations by hand to find our p-value. 

Luckily, Python has a few **libraries** which contain **functions** that allow us to conduct our hypothesis tests without doing the calculations by hand. 

Before we run these hypothesis test functions, we must **install** and `import` the library they belong to. 

To install the `statsmodels` library, try running any one of following commands in your **terminal** until successful installation:
- `python3 -m pip install statsmodels`
- `py -m pip install statsmodels`
- `pip install statsmodels`
- `pip3 install statsmodels`


Then, **run** the following `import` statement:

In [33]:
from statsmodels.stats import weightstats

This imports the `weightstats` class from the `statsmodels.stats` library, containing the function we need to perform **z-tests** in Python.

The function we will be using is `weightstats.ztest()`. This function will take **three parameters** in the following order:
1. A `list` of sampled values, where a `df["column"]` is something that can be read as a list
2. The **expected value** of the null hypothesis, specified using `value=`
3. The **alternative hypothesis** to the null value (either 'two-sided', 'larger', or 'smaller', specified using `alternative=`)

To learn more about **z-test function**, visit: https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html


## Puzzle 1.3: z-test in Python
Now, let's run the our z-test! Using the cell below, call the `weightstats.ztest()` function to conduct our z-test and store the results in the variable `sleep_results`.

Remember, our usage of `weightstats.ztest()` takes **three parameters**. Calling it will look like:

`weightstats.ztest(list, value=..., alternative=...) `

Use your knowledge about our **type of hypothesis test** and the **national sleep average** to fill in the `...`
- **IMPORTANT**: When doing the test, you must use `df["column"].dropna()` to make sure to not include any `NaN values` since this z-test requires that there are no missing values.

In [36]:
sleep_results = weightstats.ztest(df_hello[sleep_question].dropna(), value=6.8, alternative="two-sided")
sleep_results

(6.596424100776638, 4.2119297056628216e-11)

If you print `sleep_results`, it should be a pair of **two values** in the form of $(z, p)$ where $z$ is the `z-stat` and $p$ is the `p-value`. If done correctly, these values should be **almost identical** to what you calculated by hand. 

### 🔬 Test Case Checkpoint 🔬

In [37]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert('sleep_results' in vars()), "Double check your variable name for z-test result. It should be named 'sleep_results'."
assert(math.isclose(sleep_results[0], 6.596424100776638)), "The z-stat of your z-test is incorrect."
assert(math.isclose(sleep_results[1], 4.2119297056628216e-11)), "The p-value of your z-test is incorrect."
assert(math.isclose(sleep_results[0], z_stat)), "Your z-stat from sleep_result should be equivalent to the 'z_stat' variable defined earlier."
assert(math.isclose(sleep_results[1], p_value, abs_tol = 1e-15)), "Your p-value from sleep_result should be equivalent to the 'p-value' variable defined earlier."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Puzzle 1.4: How many classes are you taking this semester?
Our knowledge of the `weightstats.ztest()` function from above lets us conduct a hypothesis test on any of our columns with ease. One question you all answered on the Hello Survey was *"How many classes are you taking this semester?"*. We will answer the following question by performing a z-test:
> **Assuming the average college student is taking 5 classes in a semester, do UIUC students take *more* classes than the average college student?**

In Puzzles 1.2 and 1.3, we conducted a **two-sided z-test**, meaning we'd **reject the null** if a value was significantly **less ***or*** greater** than expected. Now, we will conduct a **one-sided z-test**, meaning we are trying to determine if something is **solely larger** or **solely smaller** than expected. 

### Defining Our Hypotheses
**Q3: Define the null and alternative hypotheses using the cell below. Remember, we assume the average college student takes 5 classes each semester and believe UIUC students may take *more*.**



Null Hypothesis: The observed difference between the average college student taking 5 classes and the number of classes a UIUC student takes is due to chance.

Alternative Hypothesis: The observed difference between the average college student taking 5 classes and the number of classes a UIUC student takes is due to something other than chance.

We can assume again that our Hello Dataset is equivalent to a **random sample** of UIUC students. Conveniently, `df_hello` contains the number of courses students are taking this semester under the **column** named `"How many classes are you taking this semester?"`.

To begin, assign this column name to a **string variable** named `course_question` using the cell below:

In [38]:
course_question = "How many classes are you taking this semester?"

## Puzzle 1.5: Conducting the Hypothesis Test

Now we have everything ready to conduct our **one-sided z-test**. 

Using the cell below, call the `weightstats.ztest()` function to conduct our **one-sided z-test**, storing the results in the variable `course_results`.

Our usage of `weightstats.ztest()` takes **three parameters**. Calling it will look like:

`weightstats.ztest(list, value=..., alternative=...) `

Remember we are conducting a **one-sided** test to determine whether students at UIUC take **more** than 5 courses per semester on average. 
- **Remember**, when doing the test, you'll want to use `df["column"].dropna()` to make sure to not include any `NaN values` since this z-test requires that there are no missing values.
- You may need to refer to the documentation to find out how to specify the correct alterative (https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html).

In [42]:
course_results = weightstats.ztest(df_hello[course_question].dropna(), value=5, alternative="larger")
course_results

(9.130709264695678, 3.402598541547406e-20)

### 🔬 Test Case Checkpoint 🔬

In [43]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert('course_results' in vars()), "Double check your variable name for z-test result. It should be named 'course_results'."
assert(math.isclose(course_results[0], 9.130709264695678)), "The z-stat of your z-test is incorrect."
assert(math.isclose(course_results[1], 3.402598541547406e-20)), "The p-value of your z-test is incorrect."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Interpretation 
**Q4: We have now observed the p-value for our hypothesis test. Do we reject the null hypothesis? Explain why or why not. Make sure to explain what this means in the context of our original question. Make sure you answer *all* parts of this question.** 

Remember, our significance level is $0.05$.



No, we do not reject the null hypothesis because p < 0.5. This means that the differene in number of classes between the average college student and the average UIUC student is not statistically significant.

<hr style="color: #DD3403;">

# Part 2: Happiness and Credit Hours
Let's do another hypothesis test! This time we are going to look at the relationship between happiness and credit hours! 

Taking random samples of student subgroups from `df_hello`, we may answer the following question:
- **Are students who take less credit hours significantly *happier* than students who take more?**

To answer, we will perform a **one-sided two-sample z-test**. Two-sample z-tests compare two samples to each other and are very common in statistics!

### Defining Our Hypotheses
**Q5: Define the null and alternative hypotheses using the cell below.**


Null Hypothesis: Any correlation between less credit hours and increased happiness is due to chance.

Alternative Hypothesis: The correllation between less credit hours and increased happiness is not due to chance.

## Puzzle 2.1: Setting Up
Now, let's run the our z-test! 

To make our code **simpler and readable**, in the next cell, create two variables for future column subselection from our `df_hello`:
1. A **string variable** named `credits_question`, containing the **string**: "How many credit hours are you taking?"
2. A **string variable** named `happiness_question`, containing the **string**:  "From 1 to 10, how happy are you right now?"


In [48]:
credits_question = "How many credit hours are you taking?"
happiness_question = "From 1 to 10, how happy are you right now?"

Again, we will need to **clean** our DataFrame before conducting these tests. Here's a list of issues we will need to take care of before doing our two sample z-test:
- Some people responded with **unreasonable** answers to the `credits_question` (any answer **above 25** does not make sense)
- Some people responded with **unreasonable** answers to the `happiness_question` (any answer **above 10** does not make sense because of the 1-10 scale)
- By checking the values for each question, these conditionals will also remove all `NaN` values!  *(Python defines `NaN` to be not greater than or less than anything, so it's always `False` and will always be removed when checked with a conditional.)*

In [49]:
df_clean = df_hello[(df_hello[credits_question] <= 25) & (df_hello[happiness_question] <= 10)]
df_clean

Name   Section  \
0        Prof. Karle    L1, L2   
2             Andrew       Y05   
3           Annabel        Y05   
4             Ayden        Y14   
5               Sens       Y30   
...              ...       ...   
1069         CA Soph  Y07, Y16   
1070          CA Ana       Y14   
1071  CA Sri Nithya        Y26   
1072       CA Lauren      YALL   
1073        CA Subin       Y05   

     What is your first name you want us to use for you?  \
0                                           Prof. Karle    
2                                                Andrew    
3                                              Annabel     
4                                                Ayden     
5                                                  Sens    
...                                                 ...    
1069                                            CA Soph    
1070                                             CA Ana    
1071                                     CA Sri Nithya     
1072                                          CA Lauren    
1073                                           CA Subin    

     What lab section are you enrolled in?  (Use the three-letter code that starts with a "Y", like "Y42".)  \
0                                                L1, L2                                                       
2                                                   Y05                                                       
3                                                   Y05                                                       
4                                                   Y14                                                       
5                                                   Y30                                                       
...                                                 ...                                                       
1069                                           Y07, Y16                                                       
1070                                                Y14                                                       
1071                                                Y26                                                       
1072                                               YALL                                                       
1073                                                Y05                                                       

     Does your NetID ending with odd number or even number?  \
0                         NetID ends with an odd number       
2                         NetID ends with an odd number       
3                        NetID ends with an even number       
4                        NetID ends with an even number       
5                        NetID ends with an even number       
...                                                 ...       
1069                     NetID ends with an even number       
1070                     NetID ends with an even number       
1071                     NetID ends with an even number       
1072                     NetID ends with an even number       
1073                     NetID ends with an even number       

     If I flipped a coin, do you think it would it land on heads or tails?  \
0                                                 Heads                      
2                                                 Heads                      
3                                                 Heads                      
4                                                 Heads                      
5                                                 Heads                      
...                                                 ...                      
1069                                              Tails                      
1070                                              Heads                      
1071                                              Heads                      
1072            

### 🔬 Test Case Checkpoint 🔬

In [50]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(df_clean[credits_question].std(), 1.8113588050790475)), "Double check the column question text you've defined in 'credits_question'." 
assert(math.isclose(df_clean[happiness_question].var(), 3.5148064721996004)), "Double check the column question text you've defined in 'happiness_question'." 
assert('df_clean' in vars()), "Your new DataFrame should be named df_clean."
assert(df_clean[credits_question].isnull().any() == False), "There are still NaN values in the credits_question column."
assert(df_clean[happiness_question].isnull().any() == False), "There are still NaN values in the happiness_question column."
assert(len(df_clean) == 1040), "Double check that you've cleaned NaN values from both question columns as well as the out-of-bounds values specified in the instructions."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


Now that we have cleaned our columns in `df_clean`, we can continue with our z-test setup. 

For this, we will consider **15 or more credit hours** as **"many"**. 

Using your `credits_question` variable, create **two** new `DataFrames`: 
- `df_many_credits`, containing the rows from `df_clean` with students taking **15 or more** credit hours. 
- `df_less_credits`, containing the rows from `df_clean` with students taking **less than 15** credit hours.

In [51]:
df_many_credits = df_clean[df_clean[credits_question] >= 15]
df_many_credits

Name Section  \
2             Andrew     Y05   
3           Annabel      Y05   
4             Ayden      Y14   
5               Sens     Y30   
6              Suvan     Y22   
...              ...     ...   
1068            Noah     Y29   
1070          CA Ana     Y14   
1071  CA Sri Nithya      Y26   
1072       CA Lauren    YALL   
1073        CA Subin     Y05   

     What is your first name you want us to use for you?  \
2                                                Andrew    
3                                              Annabel     
4                                                Ayden     
5                                                  Sens    
6                                                 Suvan    
...                                                 ...    
1068                                               Noah    
1070                                             CA Ana    
1071                                     CA Sri Nithya     
1072                                          CA Lauren    
1073                                           CA Subin    

     What lab section are you enrolled in?  (Use the three-letter code that starts with a "Y", like "Y42".)  \
2                                                   Y05                                                       
3                                                   Y05                                                       
4                                                   Y14                                                       
5                                                   Y30                                                       
6                                                   Y22                                                       
...                                                 ...                                                       
1068                                                Y29                                                       
1070                                                Y14                                                       
1071                                                Y26                                                       
1072                                               YALL                                                       
1073                                                Y05                                                       

     Does your NetID ending with odd number or even number?  \
2                         NetID ends with an odd number       
3                        NetID ends with an even number       
4                        NetID ends with an even number       
5                        NetID ends with an even number       
6                        NetID ends with an even number       
...                                                 ...       
1068                      NetID ends with an odd number       
1070                     NetID ends with an even number       
1071                     NetID ends with an even number       
1072                     NetID ends with an even number       
1073                     NetID ends with an even number       

     If I flipped a coin, do you think it would it land on heads or tails?  \
2                                                 Heads                      
3                                                 Heads                      
4                                                 Heads                      
5                                                 Heads                      
6                                                 Heads                      
...                                                 ...                      
1068                                              Tails                      
1070                                              Heads                      
1071                                              Heads                      
1072                                    

In [52]:
df_less_credits =  df_clean[df_clean[credits_question] < 15]
df_less_credits

Name   Section  \
0     Prof. Karle    L1, L2   
11          Finch       Y03   
26            Dev       Y12   
35         Thomas       Y11   
40            Kai       Y06   
...           ...       ...   
1058   TA Yinfang       Y25   
1063       Braden       Y34   
1064     Sauntrel       Y22   
1067      Nafisat       Y15   
1069      CA Soph  Y07, Y16   

     What is your first name you want us to use for you?  \
0                                           Prof. Karle    
11                                                Finch    
26                                                  Dev    
35                                               Thomas    
40                                                  Kai    
...                                                 ...    
1058                                         TA Yinfang    
1063                                             Braden    
1064                                           Sauntrel    
1067                                            Nafisat    
1069                                            CA Soph    

     What lab section are you enrolled in?  (Use the three-letter code that starts with a "Y", like "Y42".)  \
0                                                L1, L2                                                       
11                                                  Y03                                                       
26                                                  Y12                                                       
35                                                  Y11                                                       
40                                                  Y06                                                       
...                                                 ...                                                       
1058                                                Y25                                                       
1063                                                Y34                                                       
1064                                                Y22                                                       
1067                                                Y15                                                       
1069                                           Y07, Y16                                                       

     Does your NetID ending with odd number or even number?  \
0                         NetID ends with an odd number       
11                       NetID ends with an even number       
26                        NetID ends with an odd number       
35                        NetID ends with an odd number       
40                       NetID ends with an even number       
...                                                 ...       
1058                     NetID ends with an even number       
1063                     NetID ends with an even number       
1064                     NetID ends with an even number       
1067                      NetID ends with an odd number       
1069                     NetID ends with an even number       

     If I flipped a coin, do you think it would it land on heads or tails?  \
0                                                 Heads                      
11                                                Tails                      
26                                                Heads                      
35                                                Heads                      
40                                                Tails                      
...                                                 ...                      
1058                                              Heads                      
1063                                              Tails                      
1064                                              Heads                      
1067                                             

## Puzzle 2.2: Conducting the Test
Now we have the samples we need for our **two-sample z-test**.

Now, let's run the our z-test! Using the cell below, call the `weightstats.ztest()` function to conduct our z-test and store the results in the variable `happy_results`.

- Hint: See the documentation on the z-test to see how to do a two sample z-test! It's very similar to the one-sample z-tests we did in the previous puzzles!  https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

In [61]:
happy_results = weightstats.ztest(df_less_credits[credits_question], df_many_credits[credits_question], value=0, alternative="two-sided")
happy_results

(-28.806515188636485, 1.7775461647409675e-182)

### Analysis: Interpretation 
**Q6: We have now observed the p-value for our hypothesis test. Do we reject the null hypothesis? Why or why not? What does this mean in the context of our original question: Are students who take less credit hours significantly happier than students who take more? Ensure you answer *this entire* question**.

Remember, our significance level is $0.05$.


Reject the null hypotheses because 1.77 x 10^-182 < 0.05. This means that students who take less credit hours are significantly happier than students who take more.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)